README
README for mSQL version 1.0 Beta Patch 5	15 Feb 1995
----------------------------------------------------------------------------


General Info
------------

mSQL has been developed under Sun OS 4.1.1 but has been tested under
Solaris 2.3, Ultrix 4.3, Linux, and OSF/1 (cc not gcc).  That said, it 
should "autoconf" and build on most BSD derived systems, SVR4 based 
systems or POSIX O/S's (that should cover most of them).  It has been
reported that it works out-of-the-box on HP-UX, NeXT, SCO, Sequent,
Cray, Tandem and a few others.  I haven't tested these so buyer beware.


Conscienceware
--------------

mSQL is now being released as "conscienceware".  Please read the file
doc/Conscienceware for an understanding of what it is and how you can
help support the development of mSQL.


Installation
------------

To build mSQL, type "make target" in the top directory.  This will build
a directory where the object files for your system will be made.  This
mechanism allows you to NFS mount the sources onto different machines
and build mSQL on different platforms without having to make copies of 
the sources.  The directory will be under the "targets" directory (e.g.
$(TOP)/targets/Solaris-2.3-Sparc )

Once the target directory has been created, change directory into it and
run "setup".  That will run autoconf and some other configuration
scripts.  Once that has completed, just type make.  You should install
the software with "make install" when you're happy with it.  

The default installation directory is /usr/local/Minerva.  This area is
shared by all the development of Minerva (including the ESL scripting
language and Minerva itself).  It's probably a good idea to leave it as
that rather than /usr/local/msql or similar if you wish to use ESL
(check out doc/Minerva for details on ESL or grab the ESL manual via ftp
from Bond.edu.au in /pub/Minerva/esl )


Testing the software
--------------------

Once it is installed, you can try the sample mSQL script provided in
the msql source directory.  Fire up the server using

	/usr/local/Minerva/bin/msqld&

Create a test database using

	/usr/local/Minerva/bin/msqladmin create test

Run the script through the terminal monitor using

	/usr/local/Minerva/bin/msql test < sample.msql

There are also some further test scripts in the tests directory.  The
"killer" test will produce performance figures about mSQL running on
your machine.  Details such as the number of operations per second for
inserts, fills and selects on both keyed and flat tables is generated.

The most complete test is the regressions test found in test/rtest.  Run
it using 'rtest DBname' where dbname is the name of a database it can
trash.


Notes :
	o  Because this stuff is pretty new, there's a script in the
	scripts directory that you may want to use to run the server.  It
	restarts the server if it crashes and mails you a notification
	of the failure.  It's called run_daemon and may save you some 
	heart ache.

	o  If you run this on a Cray (what?  you don't have a Y-MP at
	home? :) you'll have to remove -lsocket and -lx from EXTRA_LIB 
	in site.mm after you run setup.

	o If you are building this with gcc and see link errors like
	__load.o(.text+0x1a): undefined reference to `mcount'
	edit site.mm and set 'CC=gcc -static'


If you decide to use mSQL for anything, could you drop me a line.  I'd
like to keep track of just how many people are using it and what it's
being used for.  You may also wish to join the mailing list by sending a
message to msql-list-request@Bond.edu.au.


   __                                  David J. Hughes  -  Bambi@Bond.edu.au
  /  \               /  /   /           http://Bond.edu.au/People/bambi.html
 /___/ __   _   ____/  /   / _     
/   \ /  \ / \ /   /  /   / / \  /   Senior Network Programmer, Bond University
\___/ \__//  / \__/   \__/ /  / /    Qld.   4229   AUSTRALIA    (+61 75 951450)


doc/HISTORY
mSQL 1.0 Patch 5        15 - Feb - 1995
---------------------------------------

	o Fixed bug in long text field handling
	o Added C++ support for msql.h
	o Added support for MSQL_HOME environ variable in msqld
	o Added HAVE_MMAP defines around new munmap() call
	o Extracted the install from the Makefile to an external script
	o Replaced "-c -o $obj" with "-c" in the makegen macros
	o Added partial match optimisation for the outer loop of joins
	o Fixed integer handling for Cray (64bit)


mSQL 1.0 Patch 4        8 - Feb - 1995
--------------------------------------

	o Moved target stuff to a script as bash dies on the Makefile
	o Fixed bug in net.c regarding non-terminating reads
	o Fixed memory leak left after 1.0.3 testing
	o Added identifier length checks
	o Fixed Multiple free problem with tmp tables values
	o Fixed memory leak with tmp condition list values
	o Fixed formatPacket() to handle long char fields (>1024)
	o Added hash based keyword lookup to parser  (speedup)
	o Reduced the use of bzero and bcopy (speedup)
	o Inlined a couple of functions that are called frequently (speedup)


mSQL 1.0 Patch 3        1 - Feb - 1995
--------------------------------------

	o Fixed conInfo index bug
	o Fixed net.c to ensure full reads and writes
	o Added 64-bit int support for Crays
	o Fixed Cray stuff in the README
	o Added msqlClose() to msql.h
	o Added AIX sys/select.h support to common/config.h.in
	o Added further version info to msqladmin
	o Added calls to msqlClose() in relshow
	o Added error reporting for multiple primary keys in create
	o Fixed key handling in "complex" conditions
	o Added socket details to debugging output (general in msqld
		and api in libmsql).
	o Started on primary key support withing joins
	o Reverted to an older version of autoconf so that it
		will build under SunOS 4.1.3

mSQL 1.0 Patch 2        18 - Jan - 1995
--------------------------------------

	o Fixed mmap() problem with key remapping
	o fixed -h option on msql
	o fixed tabs in # defines in site.h for AIX
	o Added "grep -i" to killer test for AIX's time output
	o Made net.c a bit more robust
	o Fixed LIKE field over-run problem
	o Added AIX stuff to sys-arch
	o Upgraded to the current version of autoconf
	o Fixed long field/table name over-runs (core dump)
	o Tightened up file and directory modes
	o Added test for max number of connections
	o Fixed socket handling on failed connection in libmsql
	o Fixed makegen macros for SVR4.2


mSQL 1.0 Patch 1        5 - Jan - 1995
--------------------------------------

	o Added -h option to msql
	o Added MSQL_HOST support to msql
	o Fixed arg count in call to msqlLoadAcl() in msqld.c
	o Fixed fchmod() problem in makedepend for SCO
	o Added ACL support for admin commands from "localhost" as
		SCO doesn't have UNIX sock's so admin commands come
		from localhost over TCP.
	o Fixed NULL value insert and condition core dumps
	o Fixed comments within char literals in the terminal monitor
	o Added -q mode for msqladmin 
	o Changed HOST to MSQLHOST in rtest to overcome default environments
	o Removed myself from run_daemon so that I don't get other
		peoples crash reports
	o Added mode setting for install directories to mkinstalldirs
	o Fixed install.mm and libinstall.mm to only chown if owned by root
	o Fixed yyerror() so that a NULL yytext doesn't cause a SEGV
	o Added quote escaping to msqldump
	o Added "integer" and "smallint" to the lexer
	o Fixed field list over-run and protocol screwup in msqlListFields()
	o Added on-the-fly INT to REAL conversion on inserts
	o Added filename and line number info to malloc and mmap debugging
	o Fixed huge memory leak in ident structure handling
	o Fixed remapping during keyed inserts (50% less map/unmap calls)
	o Fixed int active comparisons
	o Forced a buffer reload after deleteRow() if mmap() not in use.
		Thanks to Pierre Dupre (Pierre.Dupre@metro.fr) for
		finding and fixing this problem on his Motorola.
	o Fixed NULL row free in createSortedTable(), createDistinctTable()
	o Added malloc() and mmap() leak detection in debug modes
	o Added table cache cleanup on exit.
	o Added continuation prompting to monitor



mSQL 1.0                14 - Dec - 1994
---------------------------------------

	o Fixed select loop bug (bad FD on partial client connection)
	o Added read loop timeout for those with interruptable syscalls
	o Fixed one-line bug in readKey() that cause key lookups to fail
	o Merged msqlsave by Igor Romanenko (igor@frog.kiev.ua) into the
		distribution as msqldump.  Added ability to dump entire
		databases and reformated to match the rest of the code.
	o Fixed call to /bin/echo in makegen for bash
	o Added #ifdef's for signal names around calls to signal()
	o Added DISTINCT operator to SELECT's
	o Fixed bad type checking in writeKey()
	o Fixed key remapping bug (wasn't remapped between inserts).
	o Ported code to FreeBSD 2.0 (including getting around a
		compiler bug and a _very_ wierd shell!)
	o Recoded the lex scanner in C (msql_lex.c)
	o Added support for escaped data in text literals
	o Recoded text literal code to handle 8-bit data rather than
		just strings (i.e. no strlen(), strcpy() etc.)
	o Added Version info to the server
	o Added 'version' command to msqladmin
	o Added '-h host' option to all commands
	o Removed support for MSQL_HOST env var as -h does the job better
	o Added '-q' flag to msql 
	o Added regession test suite


mSQL 0.3 Beta Patch 1	28 - Nov - 1994
---------------------------------------

	o Fixed bug that caused the server to go into an endless read loop
		on a bad client crash (SIGPIPE related)
	o Added argc test to msqladmin
	o Fixed POSIX_SOURCE garbage in X's headers for makedepend
	o Recoded signal handling in makedepend/main.c so as not to
		use sigaction() and friends (just use signal() )
	o Added "proctitle" debugging option
	o Added "mmap" debugging option
	o Fixed mmap() key file seg fault after re-map
	o Generally cleanup up the mmap() based code.

src/msql/acl.c
/*
**	acl.c	- 
**
**
** Copyright (c) 1993  David J. Hughes
**
** Permission to use, copy, and distribute for non-commercial purposes,
** is hereby granted without fee, providing that the above copyright
** notice appear in all copies and that both the copyright notice and this
** permission notice appear in supporting documentation.
**
** This software is provided "as is" without any expressed or implied warranty.
**
** ID = "$Id:"
**
*/


#include <stdio.h>
#include <sys/types.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <signal.h>
#include <netdb.h>


#include <common/portability.h>
#include "msql_priv.h"


#define	ALLOW	1
#define REJECT	2

typedef	struct acc_s {
	char	name[50];
	int	access;
	struct	acc_s *next;
} acc_t;


typedef struct acl_s {
	char	db[NAME_LEN];
	acc_t	*host,
		*read,
		*write;
	tlist_t	*access,
		*option;
	struct	acl_s *next;
} acl_t;



static acl_t	*aclHead = NULL;
static int	accessPerms;

extern  char    *packet;

#define	ERR(msg)	if (verbose) printf msg 

#define	DATABASE	1
#define	READ		2
#define	WRITE		3
#define	HOST		4
#define	ACCESS		5
#define OPTION		6


static int checkToken(tok)
	char	*tok;
{
	if (strcmp(tok,"database") == 0)
		return(DATABASE);
	if (strcmp(tok,"read") == 0)
		return(READ);
	if (strcmp(tok,"write") == 0)
		return(WRITE);
	if (strcmp(tok,"host") == 0)
		return(HOST);
	if (strcmp(tok,"access") == 0)
		return(ACCESS);
	if (strcmp(tok,"option") == 0)
		return(OPTION);
	return(-1);
}



int msqlLoadAcl(verbose)
	int	verbose;
{
	acl_t	*new,
		*aclTail;
	tlist_t	*tNew,
		*tTail;
	acc_t	*accNew,
		*accTail;
	char	buf[1024],
		path[255],
		*tok;
	FILE	*fp;
	int	newEntry,
		lineNum;


	/*
	** Open the acl file
	*/
	(void)sprintf(path,"%s/msql.acl", msqlHomeDir);
	fp = fopen(path,"r");
	if (!fp)
	{
		if (verbose)
		{
			perror("Couldn't open ACL file");
		}
		sprintf(packet,"-1:Couldn't open ACL file\n");
		return(-1);
	}


	/*
	** Process the file
	*/
	fgets(buf,sizeof(buf),fp);
	newEntry = 1;
	lineNum = 1;
	while(!feof(fp))
	{
		tok = (char *)strtok(buf," \t\n\r=,");
		if (!tok)
		{
			/* Blank line ends a db entry */
			newEntry = 1;
			fgets(buf,sizeof(buf),fp);
			lineNum++;
			continue;
		}
			

		if (*tok == '#')
		{
			/* Comments are skipped */
			fgets(buf,sizeof(buf),fp);
			lineNum++;
			continue;
		}



		switch(checkToken(tok))
		{
		    case DATABASE:
			if (!newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet,
				   "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			newEntry = 0;
			tok = (char *)strtok(NULL," \t\n\r");
			if (!tok)
			{
				ERR(("Missing database name at line %d\n",
					lineNum));
				printf(packet,
					"-1:Missing database name at line %d\n",
					lineNum);
				return(-1);
			}
			new = (acl_t *)malloc(sizeof(acl_t));
			(void)bzero(new,sizeof(acl_t));
			(void)strcpy(new->db,tok);
			if (aclHead)
			{
				aclTail->next = new;
				aclTail = new;
			}
			else
			{
				aclHead = aclTail = new;
			}
			break;


		    case READ:
			if (newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet, 
				    "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			if (new->read)
			{
				accTail = new->read;
				while(accTail->next)
					accTail = accTail->next;
			}
			else
			{
				accTail = NULL;
			}
			tok = (char *)strtok(NULL," \t,\n\r");
			while(tok)
			{
				accNew = (acc_t *)malloc(sizeof(acc_t));
				(void)bzero(accNew,sizeof(acc_t));
				if (*tok == '-')
				{
					strcpy(accNew->name,tok+1);
					accNew->access = REJECT;
				}
				else
				{
					strcpy(accNew->name,tok);
					accNew->access = ALLOW;
				}
				if (accTail)
				{
					accTail->next = accNew;
				}
				else
				{
					new->read = accNew;
				}
				accTail = accNew;
				tok = (char *)strtok(NULL," \t,\n\r");
			}
			break;


		    case WRITE:
			if (newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet,
				    "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			if (new->write)
			{
				accTail = new->write;
				while(accTail->next)
					accTail = accTail->next;
			}
			else
			{
				accTail = NULL;
			}
			tok = (char *)strtok(NULL," \t,\n\r");
			while(tok)
			{
				accNew = (acc_t *)malloc(sizeof(acc_t));
				(void)bzero(accNew,sizeof(acc_t));
				if (*tok == '-')
				{
					strcpy(accNew->name,tok+1);
					accNew->access = REJECT;
				}
				else
				{
					strcpy(accNew->name,tok);
					accNew->access = ALLOW;
				}
				if (accTail)
				{
					accTail->next = accNew;
				}
				else
				{
					new->write = accNew;
				}
				accTail = accNew;
				tok = (char *)strtok(NULL," \t,\n\r");
			}
			break;


		    case HOST:
			if (newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet,
				    "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			if (new->host)
			{
				accTail = new->host;
				while(accTail->next)
					accTail = accTail->next;
			}
			else
			{
				accTail = NULL;
			}
			tok = (char *)strtok(NULL," \t,\n\r");
			while(tok)
			{
				accNew = (acc_t *)malloc(sizeof(acc_t));
				(void)bzero(accNew,sizeof(acc_t));
				if (*tok == '-')
				{
					strcpy(accNew->name,tok+1);
					accNew->access = REJECT;
				}
				else
				{
					strcpy(accNew->name,tok);
					accNew->access = ALLOW;
				}
				if (accTail)
				{
					accTail->next = accNew;
				}
				else
				{
					new->host = accNew;
				}
				accTail = accNew;
				tok = (char *)strtok(NULL," \t,\n\r");
			}
			break;


		    case ACCESS:
			if (newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet,
				    "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			if (new->access)
			{
				tTail = new->access;
				while(tTail->next)
					tTail = tTail->next;
			}
			else
			{
				tTail = NULL;
			}
			tok = (char *)strtok(NULL," \t,\n\r");
			while(tok)
			{
				tNew = (tlist_t *)malloc(sizeof(tlist_t));
				(void)bzero(tNew,sizeof(tlist_t));
				strcpy(tNew->name,tok);
				if (tTail)
				{
					tTail->next = tNew;
				}
				else
				{
					new->access = tNew;
				}
				tTail = tNew;
				tok = (char *)strtok(NULL," \t,\n\r");
			}
			break;


		    case OPTION:
			if (newEntry)
			{
				ERR(("Bad entry header location at line %d\n",
					lineNum));
				sprintf(packet,
				    "-1:Bad entry header location at line %d\n"
					,lineNum);
				return(-1);
			}
			if (new->option)
			{
				tTail = new->option;
				while(tTail->next)
					tTail = tTail->next;
			}
			else
			{
				tTail = NULL;
			}
			tok = (char *)strtok(NULL," \t,\n\r");
			while(tok)
			{
				tNew = (tlist_t *)malloc(sizeof(tlist_t));
				(void)bzero(tNew,sizeof(tlist_t));
				strcpy(tNew->name,tok);
				if (tTail)
				{
					tTail->next = tNew;
				}
				else
				{
					new->option = tNew;
				}
				tTail = tNew;
				tok = (char *)strtok(NULL," \t,\n\r");
			}
			break;


		    default:
			ERR(("Unknown ACL command \"%s\" at line %d\n", 
				tok,lineNum));
			sprintf(packet,
				"-1:Unknown ACL command \"%s\" at line %d\n", 
				tok,lineNum);
			return(-1);
			break;

		}

		fgets(buf,sizeof(buf),fp);
		lineNum++;
	}
}


dumpAcc(acc)
	acc_t	*acc;
{
	while(acc)
	{
		printf("\t\t%s %s\n",(acc->access==ALLOW)?"Yes":"No ",
			acc->name);
		acc = acc->next;
	}
}


static dumpAcl()
{
	acl_t	*curAcl;
	tlist_t	*curT;
	acc_t	*curAcc;

	curAcl = aclHead;
	while(curAcl)
	{
		printf("\nACL's for Database %s\n",curAcl->db);
		printf("\tRead access :-\n");
		dumpAcc(curAcl->read);
		printf("\tWrite access :-\n");
		dumpAcc(curAcl->write);
		printf("\tHost access :-\n");
		dumpAcc(curAcl->host);

		curAcl = curAcl->next;
	}
}



static int matchToken(pattern,tok)
	char	*pattern,
		*tok;
{
	register char 	*cp;
	char	*buf1,
		*buf2,
		*cp2;
	int	length1,
		length2;

	/*
	** Put everything to lower case
	*/
	buf1 = (char *)strdup(pattern);
	buf2 = (char *)strdup(tok);

	cp = buf1;
	while(*cp)
	{
		*cp = tolower(*cp);
		cp++;
	}

	cp = buf2;
	while(*cp)
	{
		*cp = tolower(*cp);
		cp++;
	}
	
	/*
	** Is this a wildcard?
	*/
	cp = pattern;
	if (*cp == '*')
	{
		if (*(cp+1) == 0)	/* match anything */
		{
			return(1);
		}
		length1 = strlen(cp)-1;
		length2 = strlen(tok);
		if (length1 > length2)
		{
			(void)free(buf1);
			(void)free(buf2);
			return(0);
		}
		cp2 = buf2 + length2 - length1;
		if (strcmp(cp+1,cp2) == 0)
		{
			(void)free(buf1);
			(void)free(buf2);
			return(1);
		}
		else
		{
			(void)free(buf1);
			(void)free(buf2);
			return(0);
		}
	}

	/*
	** OK, does the actual text match
	*/
	if (strcmp(buf1,buf2) == 0)
	{
		(void)free(buf1);
		(void)free(buf2);
		return(1);
	}
	else
	{
		(void)free(buf1);
		(void)free(buf2);
		return(0);
	}
}



static int matchTextList(list,tok)
	tlist_t	*list;
	char	*tok;
{
	tlist_t	*cur;

	cur = list;
	while(cur)
	{
		if (matchToken(cur->name, tok))
		{
			return(1);
		}
		cur = cur->next;
	}
	return(0);
}



static int matchAccessList(list,tok)
	acc_t	*list;
	char	*tok;
{
	acc_t	*cur;

	cur = list;
	while(cur)
	{
		if (matchToken(cur->name, tok))
		{
			if (cur->access == ALLOW)
			{
				return(1);
			}
			else
			{
				return(0);
			}
		}
		cur = cur->next;
	}
	return(0);
}



int msqlCheckAccess(db,info)
	char	*db;
	cinfo_t	*info;
{
	char 	*host,
		*user;
	struct  sockaddr *local, 
		*remote;
	acl_t	*curAcl;
	acc_t	*curAcc;
	int	res,
		perms;


	host = info->host;
	user = info->user;
	local = (struct sockaddr *)&(info->local);
	remote = (struct sockaddr *)&(info->remote);
	
	/*
	** Find an ACL entry that matches this DB
	*/
	curAcl = aclHead;
	while(curAcl)
	{
		if(matchToken(curAcl->db,db))
		{
			break;
		}
		curAcl = curAcl->next;
	}

	if (!curAcl)
	{
		return(RW_ACCESS);	/* default if no specific ACL */
	}

	/*
	** Now check the connection details
	*/
	if (!host)
	{
		/* No host == local connection */
		
		if (!matchTextList(curAcl->access,"local"))
		{
			return(NO_ACCESS);
		}
	}
	else
	{
		if (!matchTextList(curAcl->access,"remote"))
		{
			return(NO_ACCESS);
		}
		if (!matchAccessList(curAcl->host,host))
		{
			return(NO_ACCESS);
		}
	}

	/*
	** Now check the access perms
	*/
	perms = 0;
	if (matchAccessList(curAcl->read, user))
	{
		perms |= READ_ACCESS;
	}
	if (matchAccessList(curAcl->write, user))
	{
		perms |= WRITE_ACCESS;
	}
	if (perms == 0)
	{
		return(NO_ACCESS);
	}


	/*
	** Now perform any options for this connection
	*/

	
	return(perms);
}



static freeAcc(head)
	acc_t	*head;
{
	register acc_t	*curAcc,
			*prevAcc;

	curAcc = head;
	while(curAcc)
	{
		prevAcc = curAcc;
		curAcc = curAcc->next;
		(void)free(prevAcc);
	}
}

static freeTlist(head)
	tlist_t	*head;
{
	register tlist_t	*curText,
			*prevText;

	curText = head;
	while(curText)
	{
		prevText = curText;
		curText = curText->next;
		(void)free(prevText);
	}
}


static freeAcls()
{
	register acl_t	*curAcl,
			*prevAcl;

	curAcl = aclHead;
	while(curAcl)
	{
		freeAcc(curAcl->host);
		freeAcc(curAcl->read);
		freeAcc(curAcl->write);
		freeTlist(curAcl->access);
		freeTlist(curAcl->option);
		prevAcl = curAcl;
		curAcl = curAcl->next;
		(void)free(prevAcl);
	}
	aclHead = NULL;
}

reloadAcls(sock)
	int	sock;
{
	freeAcls();
	msqlLoadAcl(0);
}


msqlSetPerms(perms)
	int	perms;
{
	accessPerms = perms;
}

int msqlCheckPerms(access)
	int	access;
{
	return( accessPerms & access);
}



int msqlCheckLocal(info)
	cinfo_t	*info;
{
	char 	*host,
		*user;
	struct  sockaddr *local, 
		*remote;
	acl_t	*curAcl;
	acc_t	*curAcc;
	int	res,
		perms;


	host = info->host;
	user = info->user;
	local = (struct sockaddr *)&(info->local);
	remote = (struct sockaddr *)&(info->remote);
	if ( (!info->host || strcmp(host,"localhost") == 0) &&
	     strcmp(info->user,ROOT) == 0)
	{
		return(1);
	}
	else
	{
		return(0);
	}
}
